### PR TITLE
Add required class to labels

### PR DESCRIPTION
**Pull request by Mister-42**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  |

### CODE REVIEW

**Comment by Mister-42** [comment]
Still got an error in real web environment after tests ran perfectly.

**Comment by terabytesoftw** [comment]
Could you show how to reproduce the error ?

Thks,

**Comment by Mister-42** [comment]
```php
echo Form::widget()
    ->begin();
```

That triggered the error `Typed property Yiisoft\Form\Widget\Field::$data must not be accessed before initialization`.

**Comment by terabytesoftw** [comment]
It's weird because `Form::class` is not related to `Field::class`, add tests form empty and works, https://github.com/yiisoft/form/pull/62, it must be another code, from `Field::class` in the view.

**Comment by Mister-42** [comment]
Controller
```php
    public function weeknumbers(): ResponseInterface
    {
        return $this->viewRenderer->renderPartial('weeknumbers'); //changed to partial for test to rule out layout
    }
```

view
```php
<?php

declare(strict_types=1);

use Yiisoft\Form\Widget\Form;

echo Form::widget()->begin;
exit;
```

Results:
`Typed property Yiisoft\Form\Widget\Field::$data must not be accessed before initialization`

**Comment by Mister-42** [comment]
The issue does not occur after second commit, but I found it very strange that tests does not show this behaviour.

**Comment by terabytesoftw** [comment]
> Controlador
> 
> ```
>     public function weeknumbers(): ResponseInterface
>     {
>         return $this->viewRenderer->renderPartial('weeknumbers'); //changed to partial for test to rule out layout
>     }
> ```
> 
> Vista
> 
> ```
> <?php
> 
> declare(strict_types=1);
> 
> use Yiisoft\Form\Widget\Form;
> 
> echo Form::widget()->begin;
> exit;
> ```
> 
> Resultados:
> `Typed property Yiisoft\Form\Widget\Field::$data must not be accessed before initialization`

you are using `begin` without parentheses, its `echo Form::widget()->begin();`

**Comment by Mister-42** [comment]
Same result

**Comment by terabytesoftw** [comment]
t's weird because I literally do the same thing in the tests.

**Comment by Mister-42** [comment]
Yeah I noticed. Except there is a viewRenderer in between on the web.

**Comment by Mister-42** [comment]
- install yii-demo and run `composer u`
- replace vendor/yiisoft/form/src/Widget/Field.php with [this version](https://github.com/yiisoft/form/raw/41492af49b4a750f4963f0d3ee5c2f397fb3165c/src/Widget/Field.php)
- `vendor/bin/yii serve`
- visit http://localhost:8080/contact

There is the error

**Comment by Mister-42** [comment]
> It's weird because `Form::class` is not related to `Field::class`, add tests form empty and works, #62, it must be another code, from `Field::class` in the view.

Field is loaded by [config](https://github.com/yiisoft/yii-demo/blob/master/config/web/fields.php).

**Comment by Mister-42** [comment]
@samdark Current code works in web and in tests and can be reviewed.

**Comment by terabytesoftw** [comment]
It is not necessary to changeof the widgets to add a class, it is precisely like that, so that it supports any css framework.

```php
$fieldConfig = [
    'labelOptions()' => [['class' => 'required'],
];

Form::widget() ->begin();
    Field::Widget($fieldConfig)->config($form, 'username')
Form::end();
```

**Comment by terabytesoftw** [comment]
> > It's weird because `Form::class` is not related to `Field::class`, add tests form empty and works, #62, it must be another code, from `Field::class` in the view.
> 
> Field is loaded by [config](https://github.com/yiisoft/yii-demo/blob/master/config/web/fields.php).

You have an error in the configuration, you can place it.

**Comment by Mister-42** [comment]
> It is not necessary to change of the widgets to add a class, it is precisely like that, so that it supports any css framework.
> 
> ```
> $fieldConfig = [
>     'labelOptions()' => [['class' => 'required'],
> ];
> 
> Form::widget()->begin();
>     Field::Widget($fieldConfig)->config($form, 'username')
> Form::end();
> ```
> 
> I am against adding classes by default, only the minimum ones we have.

You want ALL fields marked as required? That is not what I am doing here...
The `required` class is very generic and can be used in any css framework. It's not even defined bootstrap css.

**Comment by Mister-42** [comment]
> > > It's weird because `Form::class` is not related to `Field::class`, add tests form empty and works, #62, it must be another code, from `Field::class` in the view.
> > 
> > 
> > Field is loaded by [config](https://github.com/yiisoft/yii-demo/blob/master/config/web/fields.php).
> 
> You have an error in the configuration, you can place it.

That's not my config, that's yii-demo

**Comment by terabytesoftw** [comment]
config global in config/fields:

```php
<?php

declare(strict_types=1);

use Yiisoft\Form\Widget\Field;

/** @var array $params */

return [
    Field::class => static function () {
        $fieldConfig = [
            'inputOptions()' => [['required' => true]],
        ];

        return Field::Widget($fieldConfig);
    }
];
```

**Comment by Mister-42** [comment]
> config global in config/fields:
> 
> ```
> <?php
> 
> declare(strict_types=1);
> 
> use Yiisoft\Form\Widget\Field;
> 
> /** @var array $params */
> 
> return [
>     Field::class => static function () {
>         $fieldConfig = [
>             'inputOptions()' => [['required' => true]],
>         ];
> 
>         return Field::Widget($fieldConfig);
>     }
> ];
> ```

That is a terrible idea. This PR adds `required` only when it is marked as required field by `Yiisoft\Validator\Rule\Required`

**Comment by terabytesoftw** [comment]
> > config global in config/fields:
> > ```
> > <?php
> > 
> > declare(strict_types=1);
> > 
> > use Yiisoft\Form\Widget\Field;
> > 
> > /** @var array $params */
> > 
> > return [
> >     Field::class => static function () {
> >         $fieldConfig = [
> >             'inputOptions()' => [['required' => true]],
> >         ];
> > 
> >         return Field::Widget($fieldConfig);
> >     }
> > ];
> > ```
> 
> That is a terrible idea. This PR adds `required` only when it is marked as required field by `Yiisoft\Validator\Rule\Required`

What sense does it make to place the validator and the required option, they are two very different validations, anyway as you have done the pr it does not work, try it.

**Comment by Mister-42** [comment]
> What sense does it make to place the validator and the required option, they are two very different validations

If it marked by validator as required, it is a required field. I can't make more sense about that than this.

> anyway as you have done the pr it does not work, try it.

Does work, see tests

**Comment by terabytesoftw** [comment]
> > What sense does it make to place the validator and the required option, they are two very different validations
> 
> If it marked by validator as required, it is a required field. I can't make more sense about that than this.
> 
> > anyway as you have done the pr it does not work, try it.
> 
> Does work, see tests

You are putting a required class in label that does nothing, must go the attribute `required` in <input required>

**Comment by Mister-42** [comment]
> Example:
> 
> ```
> <input type="text" id="contactform-username" class="form-control" name="ContactForm[username]" value="" required="" aria-required="true" placeholder="Username">
> ```

No, you got this PR all wrong

```php
<label class="control-label required" for="contactform-username">Name</label>
 <input type="text" id="contactform-username" class="form-control" name="ContactForm[username]" value="" aria-required="true" placeholder="Username">
```

**Comment by terabytesoftw** [comment]
> > Example:
> > ```
> > <input type="text" id="contactform-username" class="form-control" name="ContactForm[username]" value="" required="" aria-required="true" placeholder="Username">
> > ```
> 
> No, you got this PR all wrong
> 
> ```
> <label class="control-label required" for="contactform-username">Name</label>
>  <input type="text" id="contactform-username" class="form-control" name="ContactForm[username]" value="" aria-required="true" placeholder="Username">
> ```
> 
> > You are putting a required class in label that does nothing, must go the attribute `required` in `<input required>` , and for me it does not make sense, because the browser will validate before the yii validator.
> 
> So the label **can** by styled, giving a visual clue of that fields are required.
> 
> Example:
> 
> ```css
> .required:after {
>     color: $red;
>     content: '\a0*';
> }
> ```

I already understood I thought you wanted to validate with the browser, but you can still add your class in the general config:

in Params:

```php
    'yiisoft/form' => [
        'fieldConfig' => [
            'inputCssClass()' => ['form-control input field'],
            'labelOptions()' => [['class' => 'required', 'label' => '']],
        ],
    ],
```
config global in config/fields:

```php
<?php

declare(strict_types=1);

use Yiisoft\Form\Widget\Field;

/** @var array $params */

return [
    Field::class => static fn () => Field::Widget($params['yiisoft/form']['fieldConfig']),
];
```

**Comment by Mister-42** [comment]
> but you can still add your class in the general config:

No, because that will mark __all__ labels as required, not just the labels belonging to required fields.

**Comment by terabytesoftw** [comment]
> > but you can still add your class in the general config:
> 
> No, because that will mark **all** labels as required, not just the labels belonging to required fields.

So I already understood and if it now makes sense to me.

File: src/Widget/Field.php [hunk 1]
```diff
@@ -157,6 +157,10 @@ public function label(bool $enabledLabel = true, array $options = [], ?string $l
             $new->labelOptions['label'] = $label;
         }
 
+        if ($new->data->isAttributeRequired($new->attribute)) {
+            Html::addCssClass($options, $new->requiredCssClass);
```
Comment (samdark) [inline src/Widget/Field.php:161]: Maybe it's a better idea to add HTML5 validation attribute here? What do you think?

See https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/required

Comment (Mister-42) [reply src/Widget/Field.php:161]: That would be an attribute on the input, so browser does validation. I do class on label for visual indication for end user of required field. Different use case.

Comment (Mister-42) [reply src/Widget/Field.php:161]: ` aria-required="true"` in the input is already btw

**Comment by samdark** [comment]
👍
